File: deck/typescript/code/string/read.ts
// Read a string as whichever type it is written in.
//
// The two scanners return different shapes: `parseIpa` gives units that
// may be symbols or unknown text, `segment` gives Sounds. This flattens
// both to the one thing an encoder needs, a base with its modifiers.

import { parseIpa } from './convert'
import { segment } from './sound'
import type { Modifier, Phone } from './type'

export type ReadSound = {
  base?: Phone
  modifiers: Modifier[]
}

export function readSounds({
  text,
  type,
}: {
  text: string
  type: 'ipa' | 'tone'
}): ReadSound[] {
  if (type === 'tone') {
    return segment(text).map(sound => ({
      base: sound.base,
      modifiers: sound.modifiers,
    }))
  }

  return parseIpa(text).map(unit =>
    unit.role === 'phone'
      ? { base: unit.base, modifiers: unit.modifiers }
      : { modifiers: [] },
  )
}

File: deck/rust/code/string/read.rs
//! Read a string as whichever notation it is written in.
//!
//! The two scanners return different shapes: `parse_ipa` gives units that
//! may be symbols or unknown text, `segment` gives Sounds. This flattens
//! both to the one thing an encoder needs, a base with its modifiers.

use crate::space::model::Notation;
use crate::string::convert::{IpaUnit, parse_ipa};
use crate::string::sound::segment;
use crate::string::types::{Modifier, Phone};

pub struct ReadSound {
  pub base: Option<&'static Phone>,
  pub modifiers: Vec<&'static Modifier>,
}

pub fn read_sounds(text: &str, notation: Notation) -> Vec<ReadSound> {
  if notation == Notation::Tone {
    return segment(text)
      .into_iter()
      .map(|sound| ReadSound {
        base: sound.base,
        modifiers: sound.modifiers,
      })
      .collect();
  }

  parse_ipa(text)
    .into_iter()
    .map(|unit| match unit {
      IpaUnit::Phone { base, modifiers, .. } => {
        ReadSound { base: Some(base), modifiers }
      }
      _ => ReadSound { base: None, modifiers: Vec::new() },
    })
    .collect()
}
